src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectThreadManagementIntegrationTests.cs

```diff
@@ -0,0 +1,599 @@
+            $"Total time to process {numberOfRequestsToProcess} took {totalTimeMillis}, which exceeds " +
+            $"expected threshold of {expectedTimeThresholdToRun}");
+    }
+
+    private static int GetWorkerThreadsCurrentlyInUse()
+    {
+        ThreadPool.GetMaxThreads(out var maxWorkerThreads, out _);
+        ThreadPool.GetAvailableThreads(out var availableWorkerThreads, out _);
+        return maxWorkerThreads - availableWorkerThreads;
+    }
+
+    private WebApplicationFactory<TestStartup> SetupApp(
+        IBlobCacheService blobCacheService,
+        EventList eventList)
+    {
+        var app = _testApp.ConfigureServices(
+            services => services
+                .AddSingleton(_ => blobCacheService)
+                .AddSingleton(_ => eventList)
+                .AddMvcCore(options =>
+                {
+                    options.Filters.Add<AsyncActionFilter>();
+                    options.Filters.Add<AsyncResultFilter>();
+                }));
+
+        BlobCacheAttribute.AddService("default", blobCacheService);
+        
+        return app;
+    }
+
+    // ReSharper disable once ClassNeverInstantiated.Local
+    private class AsyncActionFilter : IAsyncActionFilter
+    {
+        private readonly EventList _eventList;
+
+        public AsyncActionFilter(EventList eventList)
+        {
+            _eventList = eventList;
+        }
+
+        public Task OnActionExecutionAsync(
+            ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            _eventList.AddEvent($"{nameof(AsyncActionFilter)} OnActionExecutionAsync called");
+        
+            var result = next();
+        
+            _eventList.AddEvent($"{nameof(AsyncActionFilter)} OnActionExecutionAsync completed - returned Task IsCompleted = {result.IsCompleted}");
+        
+            return result;
+        }
+    }
+
+    // ReSharper disable once ClassNeverInstantiated.Local
+    private class AsyncResultFilter : IAsyncResultFilter
+    {
+        private readonly EventList _eventList;
+
+        public AsyncResultFilter(EventList eventList)
+        {
+            _eventList = eventList;
+        }
+
+        public Task OnResultExecutionAsync(ResultExecutingContext context, ResultExecutionDelegate next)
+        {
+            _eventList.AddEvent($"{nameof(AsyncResultFilter)} OnResultExecutionAsync called");
+        
+            var result = next();
+        
+            _eventList.AddEvent($"{nameof(AsyncResultFilter)} OnResultExecutionAsync completed - returned Task IsCompleted = {result.IsCompleted}");
+        
+            return result;
+        }
+    }
+}
+
+/// <summary>
+/// This TestController has a single async method that has a BlobCache attribute.
+/// This method uses an artificial delay to simulate doing some async work e.g. calling some code that interacts
+/// with the database. 
+/// </summary>
+public class TestController : ControllerBase
```

Comment:
We've got a bunch of public classes here and below that will be exposed to all testing code unnecessarily. This isn't good practice and will just generally pollute code completions.

These should be marked as `internal` (but will still be exposed to the entire `Common.Tests` namespace) or should just be nested into the test class above.

Reply:
Good spot - almost surprised these weren't flagged up by Rider.  They are now nested and private.

Reply:
Remembered that the Controller needs to remain `public` and top-level in order for `AddMvc` functionality to pick it up, and therefore a bunch of its dependent classes also need to remain public, but I've nested and internalised what I can.  Interestingly `TestCacheKey` is OK to remain `internal` - I guess as we're only referring to its type rather than attempting to use it as a parameter or return type.

Reply:
Seems like it's possible to make private nested classes work: https://tpodolak.com/blog/2020/06/22/asp-net-core-adding-controllers-directly-integration-tests/

See '3. Handling private controllers'

Reply:
Ooh, nice!  Added that to our existing WebHostBuilderExtensions class and it seems to like it :+1:  Cheers!

Reply:
All done